presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -85,441 +94,708 @@
-            return (Long) value;
+            result.setColumnStatistics(columnHandle, columnStatistics);
         }
-        throw new IllegalArgumentException("Unsupported type: " + type);
+        return result.build();
     }
 
-    private OptionalDouble calculateRowsPerPartition(Map<String, PartitionStatistics> statisticsSample)
+    @VisibleForTesting
+    static OptionalDouble calculateRowsPerPartition(Collection<PartitionStatistics> statistics)
     {
-        return statisticsSample.values().stream()
+        return statistics.stream()
                 .map(PartitionStatistics::getBasicStatistics)
                 .map(HiveBasicStatistics::getRowCount)
                 .filter(OptionalLong::isPresent)
                 .mapToLong(OptionalLong::getAsLong)
+                .peek(count -> verify(count >= 0, "count must be greater than or equal to zero"))
                 .average();
     }
 
-    private Estimate calculateRowsCount(OptionalDouble rowsPerPartition, int queriedPartitionsCount)
+    private static ColumnStatistics createPartitionColumnStatistics(
+            HiveColumnHandle column,
+            Type type,
+            List<HivePartition> partitions,
+            Map<String, PartitionStatistics> statistics,
+            double rowsPerPartition,
+            double rowCount)
     {
-        if (!rowsPerPartition.isPresent()) {
-            return Estimate.unknown();
+        return ColumnStatistics.builder()
+                .setDistinctValuesCount(Estimate.of(calculateDistinctPartitionKeys(column, partitions)))
+                .setNullsFraction(Estimate.of(calculateNullsFractionForPartitioningKey(column, partitions, statistics, rowsPerPartition, rowCount)))
+                .setRange(calculateRangeForPartitioningKey(column, type, partitions))
+                .setDataSize(calculateDataSizeForPartitioningKey(column, type, partitions, statistics, rowsPerPartition))
+                .build();
+    }
+
+    @VisibleForTesting
+    static long calculateDistinctPartitionKeys(HiveColumnHandle column, List<HivePartition> partitions)
+    {
+        return partitions.stream()
+                .map(partition -> partition.getKeys().get(column))
+                .filter(value -> !value.isNull())
+                // TODO: this may not be reliable in case some of the types is not equatable
+                .distinct()
+                .count();
+    }
+
+    @VisibleForTesting
+    static double calculateNullsFractionForPartitioningKey(
+            HiveColumnHandle column,
+            List<HivePartition> partitions,
+            Map<String, PartitionStatistics> statistics,
+            double rowsPerPartition,
+            double rowCount)
+    {
+        if (rowCount == 0) {
+            return 0;
         }
-        return Estimate.of(rowsPerPartition.getAsDouble() * queriedPartitionsCount);
+        double estimatedNullsCount = partitions.stream()
+                .filter(partition -> partition.getKeys().get(column).isNull())
```

Comment:
Is this possible? I don't think partition key can be `null`.

Reply:
It is. In a partition name it is represented as `__HIVE_DEFAULT_PARTITION__`